File: tools/sdks/swiftui/Sources/ThunderSwiftUI/components/presentation/organization/OrganizationList.swift
import SwiftUI
import Thunder

/// Lists organizations the signed-in user belongs to (spec §8.4 Presentation).
public struct ThunderOrganizationList: View {
    @EnvironmentObject private var i18n: ThunderI18n
    public let onOrganizationTap: ((Organization) -> Void)?

    public init(onOrganizationTap: ((Organization) -> Void)? = nil) {
        self.onOrganizationTap = onOrganizationTap
    }

    public var body: some View {
        BaseThunderOrganizationList { orgs, isLoading, error in
            if isLoading {
                EmptyView()
            } else if orgs.isEmpty {
                Text(i18n.resolve("organizationList.empty"))
            } else {
                VStack(alignment: .leading, spacing: 0) {
                    ForEach(orgs, id: \.id) { org in
                        Button(org.name) { onOrganizationTap?(org) }
                            .accessibilityLabel(org.name)
                            .frame(minWidth: 44, minHeight: 44)
                    }
                }
            }
        }
    }
}

/// Unstyled base variant (spec §8.3).
public struct BaseThunderOrganizationList<Content: View>: View {
    @EnvironmentObject private var state: ThunderState
    public let content: ([Organization], Bool, String?) -> Content

    @State private var orgs: [Organization] = []
    @State private var isLoading = false
    @State private var error: String?

    public init(@ViewBuilder content: @escaping ([Organization], Bool, String?) -> Content) {
        self.content = content
    }

    public var body: some View {
        content(orgs, isLoading, error)
            .task { await load() }
    }

    private func load() async {
        isLoading = true
        do {
            orgs = try await state.client.getMyOrganizations()
        } catch { self.error = error.localizedDescription }
        isLoading = false
    }
}

File: tools/sdks/compose/src/main/kotlin/io/thunder/compose/components/presentation/organization/OrganizationList.kt
package io.thunder.compose.components.presentation.organization

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.BasicText
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import io.thunder.android.Organization
import io.thunder.compose.LocalThunder

/** Lists organizations the signed-in user belongs to (spec §8.4 Presentation). */
@Composable
fun ThunderOrganizationList(
    modifier: Modifier = Modifier,
    onOrganizationTap: ((Organization) -> Unit)? = null,
) {
    val state = LocalThunder.current
    val i18n = state.i18n
    BaseThunderOrganizationList(modifier = modifier) { orgs, isLoading, error ->
        when {
            isLoading -> {}
            orgs.isEmpty() -> BasicText(i18n.resolve("organizationList.empty"))
            else -> Column {
                orgs.forEach { org ->
                    BasicText(
                        org.name,
                        modifier = Modifier
                            .defaultMinSize(minHeight = 44.dp)
                            .clickable { onOrganizationTap?.invoke(org) }
                            .semantics { contentDescription = org.name },
                    )
                }
            }
        }
    }
}

/** Unstyled base variant (spec §8.3). */
@Composable
fun BaseThunderOrganizationList(
    modifier: Modifier = Modifier,
    content: @Composable (List<Organization>, Boolean, String?) -> Unit,
) {
    val state = LocalThunder.current
    var orgs by remember { mutableStateOf<List<Organization>>(emptyList()) }
    var isLoading by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(Unit) {
        isLoading = true
        try { orgs = state.client.getMyOrganizations() }
        catch (e: Exception) { error = e.message }
        isLoading = false
    }

    Box(modifier = modifier) { content(orgs, isLoading, error) }
}
